ams/server/src/main/java/com/netease/arctic/server/DefaultOptimizingService.java

```diff
@@ -84,13 +85,21 @@ public RuntimeHandlerChain getTableRuntimeHandler() {
   //TODO optimizing code
   public void loadOptimizingQueues(List<TableRuntimeMeta> tableRuntimeMetaList) {
     List<ResourceGroup> optimizerGroups = getAs(ResourceMapper.class, ResourceMapper::selectResourceGroups);
+    List<OptimizerInstance> optimizers = getAs(OptimizerMapper.class, OptimizerMapper::selectAll);
+    optimizers.forEach(optimizer -> optimizer.setTouchTime(System.currentTimeMillis()));
+    Map<String, List<OptimizerInstance>> optimizersByGroup =
+        optimizers.stream().collect(Collectors.groupingBy(OptimizerInstance::getGroupName));
     Map<String, List<TableRuntimeMeta>> groupToTableRuntimes = tableRuntimeMetaList.stream()
         .collect(Collectors.groupingBy(TableRuntimeMeta::getOptimizerGroup));
     optimizerGroups.forEach(group -> {
       String groupName = group.getName();
       List<TableRuntimeMeta> tableRuntimeMetas = groupToTableRuntimes.remove(groupName);
-      optimizingQueueByGroup.put(groupName, new OptimizingQueue(tableManager, group,
-          Optional.ofNullable(tableRuntimeMetas).orElseGet(ArrayList::new)));
+      List<OptimizerInstance> optimizersUnderGroup = optimizersByGroup.get(groupName);
+      OptimizingQueue optimizingQueue = new OptimizingQueue(tableManager, group,
+          Optional.ofNullable(tableRuntimeMetas).orElseGet(ArrayList::new),
+          Optional.ofNullable(optimizersUnderGroup).orElseGet(ArrayList::new));
+      optimizingQueueByGroup.put(groupName, optimizingQueue);
+      optimizersUnderGroup.forEach(optimizer -> optimizingQueueByToken.put(optimizer.getToken(), optimizingQueue));
```

Comment:
risk of NPE

Reply:
fix in the new commit

Reply:
Do you think all the Collections get by the mapper may return null or empty if there's no data?

Reply:
AFAIK, it returns an empty collection if there is no data.